normandy/studies/tests/test_models.py

```diff
@@ -30,30 +39,54 @@ def test_recipes_used_by(self, storage):
 
     @pytest.mark.django_db
     def test_extension_id(self, storage):
-        extension = ExtensionFactory(xpi__from_path=data_path("webext-signed.xpi"))
+        xpi = WebExtensionFileFactory(gecko_id="test-addon@normandy.mozilla.org")
+        extension = ExtensionFactory(xpi__from_func=xpi.open)
         assert extension.extension_id == "test-addon@normandy.mozilla.org"
 
     @pytest.mark.django_db
     def test_version(self, storage):
-        extension = ExtensionFactory(xpi__from_path=data_path("webext-signed.xpi"))
+        xpi = WebExtensionFileFactory(overwrite_data={"version": "0.1"})
+        extension = ExtensionFactory(xpi__from_func=xpi.open)
         assert extension.version == "0.1"
 
     @pytest.mark.django_db
     def test_hash(self, storage):
-        extension = ExtensionFactory(xpi__from_path=data_path("webext-signed.xpi"))
-        assert extension.hash == "7c0e9203de9538b1a7513d28ec344a5712f9237f2bcd494db31344dea378baf5"
+        xpi = WebExtensionFileFactory()
+        f = xpi.open()
+        hashed = hashlib.sha256(f.read()).hexdigest()
+        f.close()
+        extension = ExtensionFactory(xpi__from_func=xpi.open)
+        assert extension.hash == hashed
+
+    @pytest.mark.django_db
+    def test_no_duplicate_files(self, storage):
+        xpi = WebExtensionFileFactory()
+        ExtensionFactory(xpi__from_func=xpi.open)
+        with transaction.atomic():
+            with pytest.raises(FileExistsError):
+                ExtensionFactory(xpi__from_func=xpi.open)
```

Comment:
A minor improvement, you can combine `with`s, like this:

```python
with transaction.atomic(), pytest.raises(FileExistsError):
    ExtensionFactory(xpi__from_func=xpi.open)
```

Reply:
TIL